core/primitives/src/congestion_info.rs

```diff
@@ -112,8 +113,47 @@ impl CongestionControl {
 
     /// Whether we can accept new transaction with the receiver set to this shard.
     pub fn shard_accepts_transactions(&self) -> bool {
-        self.congestion_level() < self.config.reject_tx_congestion_threshold
+        self.rejection_reason().is_none()
     }
+
+    /// If the shard doesn't accept new transaction, provide the reason for
+    /// extra debugging information.
+    pub fn rejection_reason(&self) -> Option<RejectTransactionReason> {
+        let incoming_congestion = self.incoming_congestion();
+        let outgoing_congestion = self.outgoing_congestion();
+        let memory_congestion = self.memory_congestion();
+        let missed_chunks_congestion = self.missed_chunks_congestion();
+
+        let congestion_level = incoming_congestion
+            .max(outgoing_congestion)
+            .max(memory_congestion)
+            .max(missed_chunks_congestion);
+
+        // Convert to NotNan here, if not possible, the max above is already meaningless.
+        let congestion_level =
+            NotNan::new(congestion_level).unwrap_or_else(|_| NotNan::new(1.0).unwrap());
+        if *congestion_level < self.config.reject_tx_congestion_threshold {
+            return None;
+        }
+
+        if missed_chunks_congestion >= *congestion_level {
+            Some(RejectTransactionReason::MissedChunks { missed_chunks: self.missed_chunks_count })
+        } else if incoming_congestion >= *congestion_level {
+            Some(RejectTransactionReason::IncomingCongestion { congestion_level })
+        } else if outgoing_congestion >= *congestion_level {
+            Some(RejectTransactionReason::OutgoingCongestion { congestion_level })
+        } else {
+            Some(RejectTransactionReason::MemoryCongestion { congestion_level })
+        }
+    }
+}
+
+/// Detailed information for why a shard rejects new transactions.
+pub enum RejectTransactionReason {
+    IncomingCongestion { congestion_level: NotNan<f64> },
```

Comment:
Really cool with NotNan though I kinda dislike the NotNan name 😅

Reply:
Well, just don't read out Nan as Not-a-Number in your head and it's not all that bad :sweat_smile:

On a more serious note, I think bringing in this crate to give `Eq` to floats when necessary only makes sense. We had this problem already several times and usually make custom solutions. But this is public API for clients, hence I didn't want a custom solution. And ordered_float is in the top 300 downloaded crates. So really, we are just adopting the most common standard solution for this.